### DISCUSSION TITLE
OpenTelemetry custom SpanProcessor support

### DISCUSSION THREAD

**Discussion by amoscatelli** [quarkusio/quarkus#33349] [Q&A]
Hi guys !
Are OpenTelemetry SpanProcessor 's supported by current quarkus integration ?
When I run locally everything works when using extra/custom SpanProcessor, when I compile in native mode and deploy on kubernetes I see the SpanProcessor is correctly called, I see no error but no spans are exported.

```
import io.opentelemetry.context.Context;
import io.opentelemetry.sdk.trace.ReadWriteSpan;
import io.opentelemetry.sdk.trace.ReadableSpan;
import io.opentelemetry.sdk.trace.SpanProcessor;
import jakarta.enterprise.context.ApplicationScoped;

@ApplicationScoped
public class EndUserSpanProcessor implements SpanProcessor {

    @Override
    public void onStart(Context parentContext, ReadWriteSpan span) {
        System.out.println("HELLO!");
    }

    @Override
    public boolean isStartRequired() {
        return Boolean.TRUE;
    }

    @Override
    public void onEnd(ReadableSpan span) {
    }

    @Override
    public boolean isEndRequired() {
        return Boolean.TRUE;
    }

}

```

I also tried with this official example : 
https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/main/examples/extension/src/main/java/com/example/javaagent/DemoSpanProcessor.java

Thanks in advance

[a comment by an automated account is not shown]

**Comment by amoscatelli** [comment]
Of course if I remove the span processor, everything works again :)

**Comment by brunobat** [comment]
~You need a span exporter, not just the processor.
There are 2 ways to do this in Quarkus. Using CDI or the OTel side SPI.~

Using the SPI, you can take advantage of the Quarkus bundled `SpanExporterCDIProvider`, like in this integration test:
https://github.com/quarkusio/quarkus/blob/3e6b2a557205e7b726885792c9d616fd28042b08/integration-tests/opentelemetry-spi/src/main/resources/META-INF/services/io.opentelemetry.sdk.autoconfigure.spi.traces.ConfigurableSpanExporterProvider#L1

With the default CDI exporter, you can override the `LateBoundBatchSpanProcessor`
https://github.com/quarkusio/quarkus/blob/3e6b2a557205e7b726885792c9d616fd28042b08/extensions/opentelemetry/runtime/src/main/java/io/quarkus/opentelemetry/runtime/exporter/otlp/OtlpRecorder.java#L64

**Comment by amoscatelli** [comment]
@brunobat thank you for your answer.
Yet, I can't unrestand why I need an exporter.

I just want to modify the spans before they are exported. An Exporter would override how the spans are exported ...
According to OpenTelemetry documentation I don't need nothing more of a SpanProcessor, also this is actually working when using quarkus:dev.

Also, I would like to use the OpenTelemetry api only (and not using quarkus specific solutions).

Am I missing something ?

**Reply by brunobat** [reply]
What things do you want to modify on the spans?... There might be another way.
You can add span processors, we add them to SDK when we are building, however if you want multiple processors (yours and ours) for an exporter, you need a 'MultiSpanProcessor' and we don't support that out of the box. 
If you want to use the OTel API only, you should explore the SPI solution I mentioned before.

**Reply by amoscatelli** [reply]
I want to add attributes to the spans. To be accurate I want to add enduser.id and enduser.role based on my JWT RBAC authentication : 

```
import com.prototype.commons.Authenticated;
import io.opentelemetry.api.common.Attributes;
import io.opentelemetry.context.Context;
import io.opentelemetry.sdk.trace.ReadWriteSpan;
import io.opentelemetry.sdk.trace.ReadableSpan;
import io.opentelemetry.sdk.trace.SpanProcessor;
import io.opentelemetry.semconv.trace.attributes.SemanticAttributes;
import jakarta.enterprise.context.ApplicationScoped;
import jakarta.enterprise.context.control.ActivateRequestContext;
import jakarta.inject.Inject;
import org.apache.commons.collections4.CollectionUtils;
import org.eclipse.microprofile.context.ManagedExecutor;

//@ApplicationScoped
public class EndUserSpanProcessor implements SpanProcessor, Authenticated {

    @Inject
    protected ManagedExecutor managedExecutor;

    @Override
    @ActivateRequestContext
    public void onStart(Context parentContext, ReadWriteSpan span) {
        managedExecutor.execute(
                () -> getPrincipal().ifPresent(
                        principal -> span.setAllAttributes(
                                Attributes.of(
                                        SemanticAttributes.ENDUSER_ID,
                                        principal.getName(),
                                        SemanticAttributes.ENDUSER_ROLE,
                                        CollectionUtils.emptyIfNull(
                                                principal.getGroups()
                                        ).toString()
                                )
                        )
                )
        );
    }

    @Override
    public boolean isStartRequired() {
        return Boolean.TRUE;
    }

    @Override
    public void onEnd(ReadableSpan span) {
    }

    @Override
    public boolean isEndRequired() {
        return Boolean.FALSE;
    }

}

```

Of course ApplicationScoped is commented because in native mode, as I reported, this prevented my spans to be exported.
Before you ask, I am using ManagedExecutor because otherwise vert-x complains about being blocked.
Again, everything works in dev mode.

**Reply by brunobat** [reply]
I didn't get this was working on JVM but not in native mode.
Do you get any error?

**Reply by amoscatelli** [reply]
> I didn't get this was working on JVM but not in native mode. Do you get any error?

Sadly no, I see no errors.
In native mode no spans are exported, and no errors are showed, as soon as I deploy with an ApplicationScoped SpanProcessor, even if it is empty and does nothing.

I think my code/attempt is legit, because of this (please have a look) : 
https://github.com/quarkusio/quarkus/blob/67644b6b5f33ec4ddd6e2c801fc5e977e2b8fab4/extensions/opentelemetry/runtime/src/main/java/io/quarkus/opentelemetry/runtime/OpenTelemetryProducer.java#L55

As far as I understand this, Quarkus looks for and inject EVERY SpanProcessor and add them to the tracer.
Did I get this wrong ?

Infact in jvm mode the SpanProcessor is correctly injected and called, but only in native mode this breaks everything (again without error).

**Reply by brunobat** [reply]
Correct, the processos should be added there.
Can you please file a bug?

**Reply by amoscatelli** [reply]
Sure thing, can you try to reproduce this too ?
As I already said, even an empty SpanProcessor should be enough, so it should be easy to reproduce

**Reply by brunobat** [reply]
What you are trying to do is super handy. 
Once we sort this out, you could submit a PR to add that functionality to the main extension if auth data is available and a flag is set to do the output.

**Reply by amoscatelli** [reply]
Yea that was my intention to, I am happy you consider this handy.
To tell the truth opentelemetry java agent already does this automatically for other runtimes (for example Wildfly/Undertow).
So yes, I think this should be brought to the main extension too.

**Reply by amoscatelli** [reply]
I created the issue

**Reply by brunobat** [reply]
Thanks for https://github.com/quarkusio/quarkus/issues/33407

**Reply by amoscatelli** [reply]
@brunobat the issue is solved, so maybe we could do as you said (add the functionality to the main extension)

**Reply by brunobat** [reply]
Sure. Can you please open an issue with the details of the behaviour you want?
We can then discuss there all the details.
